src/profile-logic/profile-data.js

```diff
@@ -2549,39 +2548,32 @@ export function filterToDeallocationsMemory(
     const bytes = nativeAllocations.weight[allocationIndex];
     const memoryAddress = nativeAllocations.memoryAddress[allocationIndex];
     if (bytes >= 0) {
-      // Handle the allocation.
+      // This is an allocation.
 
       // Provide a map back to this index.
-      memoryAddressToAllocation.set(memoryAddress, allocationIndex);
-      stackOfOriginalAllocation[allocationIndex] = null;
-    } else {
-      // Lookup the previous allocation.
-      const previousAllocationIndex = memoryAddressToAllocation.get(
-        memoryAddress
-      );
-      if (previousAllocationIndex === undefined) {
-        stackOfOriginalAllocation[allocationIndex] = null;
-      } else {
-        // This deallocation matches a previous allocation. Remove the allocation.
-        stackOfOriginalAllocation[allocationIndex] = previousAllocationIndex;
-        // There is a match, so delete this old association.
-        memoryAddressToAllocation.delete(memoryAddress);
-      }
+      memoryAddressToAllocationSite[memoryAddress] =
+        nativeAllocations.stack[allocationIndex];
+      continue;
     }
-  }
 
-  const newDeallocations = getEmptyBalancedNativeAllocationsTable();
-  for (let i = 0; i < nativeAllocations.length; i++) {
-    const duration = nativeAllocations.weight[i];
-    const stackIndex = stackOfOriginalAllocation[i];
-    if (stackIndex !== null) {
-      newDeallocations.time.push(nativeAllocations.time[i]);
-      newDeallocations.stack.push(stackIndex);
-      newDeallocations.weight.push(duration);
-      newDeallocations.memoryAddress.push(nativeAllocations.memoryAddress[i]);
-      newDeallocations.threadId.push(nativeAllocations.threadId[i]);
-      newDeallocations.length++;
+    // This is a deallocation.
+    // Lookup the previous allocation.
+    const allocationStackIndex = memoryAddressToAllocationSite[memoryAddress];
+    if (allocationStackIndex === undefined) {
+      // This deallocation doesn't match an allocation. Let's bail out.
+      continue;
     }
+
+    // This deallocation matches a previous allocation.
+    newDeallocations.time.push(nativeAllocations.time[allocationIndex]);
+    newDeallocations.stack.push(allocationStackIndex);
+    newDeallocations.weight.push(bytes);
+    newDeallocations.memoryAddress.push(memoryAddress);
+    newDeallocations.threadId.push(nativeAllocations.threadId[allocationIndex]);
+    newDeallocations.length++;
+
+    // Remove the saved allocation
```

Comment:
Thought: I was somewhat confused as to why this is deleting here, I see that my original implementation also did it. I believe the memory sampling operation is somewhat fallible, so it's probably good to account for the failure mode. I wonder if it's worth elaborating about this in the comment. If the sampling were 100% correct we wouldn't need this line, as every allocation and de-allocation would be accounted for.

Reply:
I initially didn't include it in my first implementation. Then I realized that it's (at least theorically) possible that several allocations happen at the same memory address... I mean after one is frees, I believe it's possible it's allocated again. That's why I added back the deletion line (with a test for this case!).

I'm not sure I covered all possible cases around this though. Also I don't know how reallocs work. Are we tracking them as a dealloc + alloc?

Reply:
The _allocations_ are sampled, the _deallocations_ are deterministically gathered, however in a fallible manner. So a sampled allocation will always _attempt_ to track the deallocation, but it may fail to actually record it. This is what balanced means here. The actual deallocations that are failed to be collected should be quite low, and be an anomaly.

Reply:
The case I'm mentioning here is this:
1. allocation happens at address A, it's collected
2. A is deallocated, it's collected too
3. after some time, another allocation happens at address A and it's collected too
4. A is deallocated again

=> We want that the second deallocation be tied to the to second allocation.

This makes me think that if we don't "delete" the first allocation, it would still be overwritten by the second allocation at step 3 with this algorithm. Then I wonder if we shouldn't at least output something in the console, if `memoryAddressToAllocationSite[memoryAddress]` already exists when we want to write to it... (not Error for real, because as you wrote, this can happen in the wild and still be valid).

About reallocations: I'm thinking of things like https://searchfox.org/mozilla-central/rev/6e630edb09c3ab06d0103665b16c9ea7dce782c5/memory/build/mozjemalloc.cpp#4101, is that instrumented?